core/src/util/shaderProgram.cpp

```diff
@@ -278,6 +279,24 @@ void ShaderProgram::invalidateAllPrograms() {
     
 }
 
+void ShaderProgram::setUniform(const std::string& _name, Texture& _texture) {
+    use();
+    GLint location = getUniformLocation(_name);
+    GLuint unit;
+    
+    if (m_textureSlots.find(_name) == m_textureSlots.end()) {
+        unit = m_freeTextureUnit;
+        m_textureSlots[_name] = { _texture.getGlHandle(), m_freeTextureUnit++ };
+    } else {
+        unit = m_textureSlots[_name].second;
+    }
+    
+    _texture.update(unit);
+    _texture.bind(unit);
+    
+    glUniform1i(location, unit);
+}
+
```

Comment:
I was actually thinking we could just take out this higher-level uniform logic and manually apply uniforms (probably more efficient than any algorithm, since we can use context-specific information)